packages/compiler/src/shadow_css.ts

```diff
@@ -157,6 +172,115 @@ export class ShadowCss {
     return this._insertPolyfillRulesInCssText(cssText);
   }
 
+  /*
+   * Process styles to add scope to keyframes, this implies modifying both
+   * the names of the keyframes defined in the component styles and also the css
+   * animation rules using them (animation rules using keyframes defined
+   * elsewhere need not to be modified to allow for globally defined keyframes)
+   *
+   * For example, we convert this css:
+   *
+   * .box {
+   *   animation: box-animation 1s forwards;
+   * }
+   *
+   * @keyframes box-animation {
+   *   to {
+   *     background-color: green;
+   *   }
+   * }
+   *
+   * to this:
+   *
+   * .box {
+   *   animation: scopeName_box-animation 1s forwards;
+   * }
+   *
+   * @keyframes scopeName_box-animation {
+   *   to {
+   *     background-color: green;
+   *   }
+   * }
+   *
+   **/
+  private _scopeLocalKeyframes(cssText: string, scopeSelector: string): string {
+    // This set is used to collect all keyframes names (before scoping) so that they can
+    // later be used to scope the animation rules.
+    const unscopedKeyframes = new Set<string>();
+
+    const COMMA_IN_PLACEHOLDER = '%COMMAIN%';
+
+    const str = (input?: string) => input || '';
+    const hasQuotes = (input: string) => input.includes('"') || input.includes('\'');
+
+    const scopedKeyframesCssText = processRules(cssText, rule => {
+      let selector = rule.selector;
+      if (hasQuotes(selector)) {
+        selector = selector.replace(
+            /^@(-webkit-)?keyframes(\s+)(['"])([^'"]+)\3\s*$/,
+            (_, webkit, spaces, quote, keyframeName) => {
+              unscopedKeyframes.add(keyframeName);
+              return `@${str(webkit)}keyframes${spaces}${str(quote)}${scopeSelector}_${
+                  keyframeName}${str(quote)}`;
```

Comment:
We can avoid the need for `str()` here: either by providing an initializer or noting that for the string to match the group must exist:

```suggestion
            (_, webkit = '', spaces, quote, keyframeName) => {
              unscopedKeyframes.add(keyframeName);
              return `@${webkit}keyframes${spaces}${quote}${scopeSelector}_${
                  keyframeName}${quote}`;
```

Reply:
See [playground](https://www.typescriptlang.org/play?#code/MYewdgzgLgBBCmAbewogE4wLwwOQAEBreATwDN0BDAW3ghgB1cyQQncBuAKC9EljDwA7gGUkKNJhwJkqDADp08AA6JKweAAouMXXv26A9AD18mgLRD4AI0IBLKOYCUAfmLkqtCJoYQA1E6aANq4AEQAuoFBxmHhAQwAzL4AVAAkhgA0OgY5mgD6GTBWtg7YeLiFEMrqdIUAjgCuIFDwhe4UNPAAcp1O2AB8MADe2Tk5fBAgyPKIIADm+U7cY2MTU-Az85rF9lBLoyt6a9OzC1U1EPuHBscbp5qNzfBX10fgkydb7Z7dvcuvuiUUAa6DA5X+1wAvksgA)

Reply:
yeah sorry, I got so focused on the other issues that I missed the simplest thing :cry: :sob:

`str` is 100% useless, I'll get rid of it, sorry :disappointed: :bow: